src/api_client/api_client.py

```diff
@@ -20,18 +24,91 @@ def get_exchange_rates(self, currency_pair, interval, start_time, end_time):
+        portfolio = portfolio.set_index('asset')
+        portfolio = portfolio.astype(float)
+        return portfolio[portfolio['free'] > 0]
+
+    def get_commission(self, currency_pair=f'{config.CURRENCY_TO_BUY}{config.CURRENCY_TO_SELL}'):
+        fees = self.client.get_trade_fee(symbol=currency_pair)
+        return float(fees[0]['takerCommission'])
+
+    def get_amount_precision(self, currency_pair=f'{config.CURRENCY_TO_BUY}{config.CURRENCY_TO_SELL}'):
+        asset_info = self.client.get_symbol_info(currency_pair)
+        step_size = float(asset_info['filters'][2]['stepSize'])
+        return int(np.log10(round(1 / step_size)))
+
+    def get_minimum_amount(self, currency_to_buy, currency_to_sell):
+        asset_info = self.client.get_symbol_info(f'{currency_to_buy}{currency_to_sell}')
+        if asset_info['filters'][3]['applyToMarket']:
+            amount_usdt = float(asset_info['filters'][3]['minNotional'])
+        else:
+            amount_usdt = float(asset_info['filters'][5]['minNotional'])
+        price = self.get_current_exchange_rate(currency_to_buy=currency_to_buy, currency_to_sell=currency_to_sell)
+        return amount_usdt / price
+
+
+class BinanceBackTestClient(BinanceClient):
+    def __init__(self, api_key, api_secret, **kwargs):
+        super().__init__(api_key, api_secret, **kwargs)
+        self.balances = pd.DataFrame(
+            data=[[0.02, 0],   # BTC
+                  [1000, 0]],  # EUR
+            index=[config.CURRENCY_TO_BUY, config.CURRENCY_TO_SELL],
+            columns=['free', 'locked']
+        )
+        self.commission = super().get_commission(f'{config.CURRENCY_TO_BUY}{config.CURRENCY_TO_SELL}')
+        self.precision = super().get_amount_precision(f'{config.CURRENCY_TO_BUY}{config.CURRENCY_TO_SELL}')
+        self.min_amount = 0.00022237417791046103  # super().get_minimum_amount(config.CURRENCY_TO_BUY, config.CURRENCY_TO_SELL)
+
+    def get_current_exchange_rate(self, currency_to_buy, currency_to_sell, data=None):
+        if f"{currency_to_buy}{currency_to_sell}_Open" in data.columns:
+            return data.iloc[-1][f"{currency_to_buy}{currency_to_sell}_Open"]
+        else:
+            return 1 / data.iloc[-1][f"{currency_to_sell}{currency_to_buy}_Open"]
+
+    def get_balances(self):
+        return self.balances
+
+    def get_commission(self, *args, **kwargs):
+        return self.commission
+
+    def get_amount_precision(self, *args, **kwargs):
+        return self.precision
+
+    def get_minimum_amount(self, *args, **kwargs):
+        return self.min_amount
 
-    def view_order(self, order_id):
-        pass
+    def trade(self, currency_to_buy, currency_to_sell, action, amount, data=None):
+        exchange_rate = self.get_current_exchange_rate(currency_to_buy, currency_to_sell, data)
+        commission = self.get_commission()
+        if action == enums.SIDE_BUY:
+            self.balances.at[currency_to_sell, 'free'] -= amount * exchange_rate * (1 + commission)
```

Comment:
Are you checking somewhere that amount + comission is not over the available value?

Reply:
Yep, [here](https://github.com/pavlo-seimskyi/crypto-trading-bot/blob/adding-backtesting-module/src/order_executer/service/portfolio.py#L95)